src/handlers/swap_quote_handlers.ts

```diff
@@ -0,0 +1,199 @@
+import { BigNumber, RevertError } from '@0x/utils';
+import { TxData, Web3Wrapper } from '@0x/web3-wrapper';
+import * as express from 'express';
+import * as HttpStatus from 'http-status-codes';
+
+import { CHAIN_ID } from '../config';
+import {
+    DEFAULT_QUOTE_SLIPPAGE_PERCENTAGE,
+    DEFAULT_TOKEN_DECIMALS,
+    ETH_SYMBOL,
+    QUOTE_ORDER_EXPIRATION_BUFFER_MS,
+} from '../constants';
+import { InternalServerError, RevertAPIError, ValidationError, ValidationErrorCodes } from '../errors';
+import { isAPIError } from '../middleware/error_handling';
+import { GetSwapQuoteRequestParams, GetSwapQuoteResponse } from '../types';
+import { findTokenAddress, findTokenDecimalsIfExists } from '../utils/token_metadata_utils';
+
+export class SwapQuoteHandlers {
+    private readonly _provider: SupportedProvider;
+    private readonly _swapQuoter: SwapQuoter;
+    private readonly _swapQuoteConsumer: SwapQuoteConsumer;
+    private readonly _web3Wrapper: Web3Wrapper;
+    constructor(orderbook: Orderbook, provider: SupportedProvider) {
+        this._provider = provider;
+        const swapQuoterOpts = {
+            chainId: CHAIN_ID,
+            expiryBufferMs: QUOTE_ORDER_EXPIRATION_BUFFER_MS,
+        };
+        this._swapQuoter = new SwapQuoter(this._provider, orderbook, swapQuoterOpts);
+        this._swapQuoteConsumer = new SwapQuoteConsumer(this._provider);
+        this._web3Wrapper = new Web3Wrapper(this._provider);
+    }
+    public async calculateSwapQuoteAsync(
+        buyTokenAddress: string,
+        sellTokenAddress: string,
+        buyAmount: BigNumber | undefined,
+        sellAmount: BigNumber | undefined,
+        from: string | undefined,
+        isETHSell: boolean = false,
+        slippagePercentage?: number,
+        providedGasPrice?: BigNumber,
+    ): Promise<GetSwapQuoteResponse> {
+        let swapQuote;
+        if (sellAmount !== undefined) {
+            swapQuote = await this._swapQuoter.getMarketSellSwapQuoteAsync(
+                buyTokenAddress,
+                sellTokenAddress,
+                sellAmount,
+                { slippagePercentage, gasPrice: providedGasPrice },
+            );
+        } else if (buyAmount !== undefined) {
+            swapQuote = await this._swapQuoter.getMarketBuySwapQuoteAsync(
+                buyTokenAddress,
+                sellTokenAddress,
+                buyAmount,
+                { slippagePercentage, gasPrice: providedGasPrice },
+            );
+        } else {
+            throw new Error('sellAmount or buyAmount required');
+        }
+        const {
+            makerAssetAmount,
+            totalTakerAssetAmount,
+            protocolFeeInWeiAmount: protocolFee,
+        } = swapQuote.worstCaseQuoteInfo;
+        const { orders, gasPrice } = swapQuote;
+
+        // If ETH was specified as the token to sell then we use the Forwarder
+        const extensionContractType = isETHSell ? ExtensionContractType.Forwarder : ExtensionContractType.None;
```

Comment:
This kind of automatic picking has gotten us into trouble in the past. Should we allow the request to (optionally) set the extension contract somehow?

Reply:
They can control it by choosing `ETH` (Forwarder) or `WETH` (Exchange) as the token, we're not choosing for them based on what they have available.